sw/device/lib/dif/dif_rv_timer.h

```diff
@@ -0,0 +1,302 @@
+   * hertz. In particular,
+   *
+   *   prescale = (clock_freq / tick_freq) - 1
+   *
+   * This value is declared as a uint16_t, but only the lowest 12 bits are
+   * actually used.
+   */
+  uint16_t prescale;
+
+  /**
+   * The amount to increment the timer counter at each tick.
+   */
+  uint8_t tick_step;
+} dif_rv_timer_tick_params_t;
+
+/**
+ * Errors returned by |dif_rv_timer_init()|.
+ */
+typedef enum dif_rv_timer_approximate_tick_params_result {
+  /**
+   * Indicates a successful operation.
+   */
+  kDifRvTimerApproximateTickParamsOk = 0,
+  /**
+   * Indicates a failed precondition on the function arguments.
+   */
+  kDifRvTimerApproximateTickParamsBadArg,
+  /**
+   * Indicates that the requested counter frequency cannot be represented
+   * without loss in precission.
+   */
+  kDifRvTimerApproximateTickParamsUnrepresentable,
+} dif_rv_timer_approximate_tick_params_result_t;
+
+/**
+ * Generates an aproximate |dif_rv_timer_tick_params_t| given the device
+ * clock frequency and desired counter frequency (both given in Hertz).
+ *
+ * For the purposes of this function, "counter frequency" is the frequency
+ * at which software would observe a timer counter to increase. If the
+ * clock has insufficient resolution, high counter frequencies may set a
+ * larger value for |tick_step|. For example, if the clock ticks at 50kHz,
+ * but we want a counter that seems to tick every microsecond (1MHz),
+ * we can achieve this with a prescale of 0 (so that there is a tick per
+ * clock cycle) and a tick step of 20 (since 20 * 50kHz = 1MHz).
+ *
+ * The return value of this function is only an approximation, and the
+ * actual counter frequency ultimately depends on the accuracy of the
+ * clock.
+ *
+ * @param clock_freq the device clock frequency, in Hertz.
+ * @param counter_freq the desired counter frequency, in Hertz.
+ * @param out tick parameters that will approximately produce the desired
+ *         counter frequency.
+ * @return the result of the operation.
+ */
+dif_rv_timer_approximate_tick_params_result_t
+dif_rv_timer_approximate_tick_params(uint64_t clock_freq, uint64_t counter_freq,
+                                     dif_rv_timer_tick_params_t *out);
+
+/**
+ * Configuration for initializing a RISC-V timer.
+ */
+typedef struct dif_rv_timer_config {
+  /**
+   * The physical location of the timer in memory.
+   */
+  mmio_region_t base_addr;
+
+  /**
+   * Timekeeping parameters to initialize with.
+   */
+  dif_rv_timer_tick_params_t tick_params;
+
+  /**
+   * The index of the RISC-V hardware thread that owns this timer.
+   */
+  uint32_t hart;
+} dif_rv_timer_config_t;
+
+/**
+ * State for a particular RISC-V timer, associated with a hardware thread.
+ *
+ * Its member variables should be considered private, and are only provided so
+ * that callers can allocate it.
+ */
+typedef struct dif_rv_timer {
+  mmio_region_t base_addr;
+  uint32_t hart;
```

Comment:
I think this doesn't belong in the `dif_rv_timer_t`, as it implies that these states need to be created for each HART, resulting in a number of `init` calls. I had the same issue with PLIC, where it was nice to keep some data in `dif_plic_t` , but I ended up passing it directly in all the API calls that required this variable.

Reply:
This has been addressed in the new API.